File: src/main/scala/werger/adapters/googletranslate/GoogleTranslateSource.scala
package werger.adapters.googletranslate

import cats.effect.IO
import org.http4s.client.UnexpectedStatus
import werger.domain.Language
import werger.ports.{MachineTranslationSource, MtError}

object GoogleTranslateSource:
  // Google Translate has covered Kurdish under the single code "ku" since adding it in 2016; verify against
  // Google's current supported-languages list before depending on this in production.
  val kurmanjiOnly: Map[String, String] = Map("kmr" -> "ku")

class GoogleTranslateSource(
    client: GoogleTranslateClient,
    languageCodes: Map[String, String] = GoogleTranslateSource.kurmanjiOnly
) extends MachineTranslationSource:
  def translateBatch(language: Language, sourceTexts: List[String]): IO[Either[MtError, Map[String, String]]] =
    languageCodes.get(language.code) match
      case None =>
        IO.pure(Left(MtError.Unsupported(s"No Google Translate code mapped for language ${language.code}")))
      case Some(providerCode) =>
        client
          .translate(sourceTexts, providerCode)
          .map(Right(_))
          .recover {
            case UnexpectedStatus(status, _, _) if status.code == 400 =>
              Left(MtError.Unsupported(s"Google Translate rejected the request: $status"))
            case UnexpectedStatus(status, _, _) =>
              Left(MtError.Transient(s"Google Translate returned $status"))
            case e: Throwable =>
              Left(MtError.Transient(e.getMessage))
          }

File: src/main/java/werger/adapters/googletranslate/GoogleTranslateSource.java
package werger.adapters.googletranslate;

import java.util.List;
import java.util.Map;
import org.springframework.web.client.RestClientResponseException;
import werger.domain.Language;
import werger.ports.MachineTranslationSource;
import werger.ports.MtError;
import werger.ports.Result;

public class GoogleTranslateSource implements MachineTranslationSource {
    // Google Translate has covered Kurdish under the single code "ku" since adding it in 2016; verify against
    // Google's current supported-languages list before depending on this in production.
    public static final Map<String, String> KURMANJI_ONLY = Map.of("kmr", "ku");

    private final GoogleTranslateClient client;
    private final Map<String, String> languageCodes;

    public GoogleTranslateSource(GoogleTranslateClient client) {
        this(client, KURMANJI_ONLY);
    }

    public GoogleTranslateSource(GoogleTranslateClient client, Map<String, String> languageCodes) {
        this.client = client;
        this.languageCodes = languageCodes;
    }

    @Override
    public Result<MtError, Map<String, String>> translateBatch(Language language, List<String> sourceTexts) {
        String providerCode = languageCodes.get(language.code());
        if (providerCode == null) {
            return Result.err(
                    new MtError.Unsupported("No Google Translate code mapped for language " + language.code()));
        }
        try {
            return Result.ok(client.translate(sourceTexts, providerCode));
        } catch (RestClientResponseException e) {
            return Result.err(
                    e.getStatusCode().value() == 400
                            ? new MtError.Unsupported("Google Translate rejected the request: " + e.getStatusCode())
                            : new MtError.Transient("Google Translate returned " + e.getStatusCode()));
        } catch (UnexpectedResponseShapeException e) {
            return Result.err(new MtError.Unsupported(
                    "Google Translate response didn't match the expected shape: " + e.getMessage()));
        } catch (RuntimeException e) {
            return Result.err(new MtError.Transient(String.valueOf(e.getMessage())));
        }
    }
}
